# Importing Libs

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns

In [ ]:
# Link do dataset (não consegui subir aqui no colab)
# https://drive.google.com/file/d/1jH77Zsu6ulfIkKm6ydU5p8napoW0pAJA/view?usp=sharing

In [ ]:
dataset = pd.read_csv('/content/adult.csv')
dataset.describe()

age        fnlwgt  education.num  capital.gain  capital.loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours.per.week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

# Dataset understading

In [ ]:
dataset.head(10)

age    workclass  fnlwgt     education  education.num marital.status  \
0   90            ?   77053       HS-grad              9        Widowed   
1   82      Private  132870       HS-grad              9        Widowed   
2   66            ?  186061  Some-college             10        Widowed   
3   54      Private  140359       7th-8th              4       Divorced   
4   41      Private  264663  Some-college             10      Separated   
5   34      Private  216864       HS-grad              9       Divorced   
6   38      Private  150601          10th              6      Separated   
7   74    State-gov   88638     Doctorate             16  Never-married   
8   68  Federal-gov  422013       HS-grad              9       Divorced   
9   41      Private   70037  Some-college             10  Never-married   

          occupation    relationship   race     sex  capital.gain  \
0                  ?   Not-in-family  White  Female             0   
1    Exec-managerial   Not-in-family  White  Female             0   
2                  ?       Unmarried  Black  Female             0   
3  Machine-op-inspct       Unmarried  White  Female             0   
4     Prof-specialty       Own-child  White  Female             0   
5      Other-service       Unmarried  White  Female             0   
6       Adm-clerical       Unmarried  White    Male             0   
7     Prof-specialty  Other-relative  White  Female             0   
8     Prof-specialty   Not-in-family  White  Female             0   
9       Craft-repair       Unmarried  White    Male             0   

   capital.loss  hours.per.week native.country income  
0          4356              40  United-States  <=50K  
1          4356              18  United-States  <=50K  
2          4356              40  United-States  <=50K  
3          3900              40  United-States  <=50K  
4          3900              40  United-States  <=50K  
5          3770              45  United-States  <=50K  
6          3770              40  United-States  <=50K  
7          3683              20  United-States   >50K  
8          3683              40  United-States  <=50K  
9          3004              60              ?   >50K

In [ ]:
dataset.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education.num',
       'marital.status', 'occupation', 'relationship', 'race', 'sex',
       'capital.gain', 'capital.loss', 'hours.per.week', 'native.country',
       'income'],
      dtype='object')

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [ ]:
features = ['age', 'workclass', 'fnlwgt', 'education', 'education.num',
       'marital.status', 'occupation', 'relationship', 'race', 'sex',
       'capital.gain', 'capital.loss', 'hours.per.week', 'native.country']

In [ ]:
X = dataset[features].values
X

array([[90, '?', 77053, ..., 4356, 40, 'United-States'],
       [82, 'Private', 132870, ..., 4356, 18, 'United-States'],
       [66, '?', 186061, ..., 4356, 40, 'United-States'],
       ...,
       [40, 'Private', 154374, ..., 0, 40, 'United-States'],
       [58, 'Private', 151910, ..., 0, 40, 'United-States'],
       [22, 'Private', 201490, ..., 0, 20, 'United-States']], dtype=object)

In [ ]:
X

array([[90, '?', 77053, ..., 4356, 40, 'United-States'],
       [82, 'Private', 132870, ..., 4356, 18, 'United-States'],
       [66, '?', 186061, ..., 4356, 40, 'United-States'],
       ...,
       [40, 'Private', 154374, ..., 0, 40, 'United-States'],
       [58, 'Private', 151910, ..., 0, 40, 'United-States'],
       [22, 'Private', 201490, ..., 0, 20, 'United-States']], dtype=object)

## Enconding data

In [ ]:
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()

In [ ]:
X[:, 1] = label_encoder.fit_transform(X[:, 1])
X[:, 3] = label_encoder.fit_transform(X[:, 3])
X[:, 5] = label_encoder.fit_transform(X[:, 5])
X[:, 6] = label_encoder.fit_transform(X[:, 6])
X[:, 7] = label_encoder.fit_transform(X[:, 7])
X[:, 8] = label_encoder.fit_transform(X[:, 8])
X[:, 9] = label_encoder.fit_transform(X[:, 9])
X[:, 13] = label_encoder.fit_transform(X[:, 13])

In [ ]:
dataset.head(5)

age workclass  fnlwgt     education  education.num marital.status  \
0   90         ?   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66         ?  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital.gain  \
0                  ?  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2                  ?      Unmarried  Black  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   

   capital.loss  hours.per.week native.country income  
0          4356              40  United-States  <=50K  
1          4356              18  United-States  <=50K  
2          4356              40  United-States  <=50K  
3          3900              40  United-States  <=50K  
4          3900              40  United-States  <=50K

In [ ]:
X

array([[90, 0, 77053, ..., 4356, 40, 39],
       [82, 4, 132870, ..., 4356, 18, 39],
       [66, 0, 186061, ..., 4356, 40, 39],
       ...,
       [40, 4, 154374, ..., 0, 40, 39],
       [58, 4, 151910, ..., 0, 40, 39],
       [22, 4, 201490, ..., 0, 20, 39]], dtype=object)

## Converting Label to a numeric format

In [ ]:
lE = LabelEncoder()
y = lE.fit_transform(dataset['income'])

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [ ]:
scaler = StandardScaler()

In [ ]:
X = scaler.fit_transform(X)
X

array([[ 3.76961234, -2.65732045, -1.06799736, ..., 10.59350656,
        -0.03542945,  0.29156857],
       [ 3.18311167,  0.09005041, -0.53916866, ..., 10.59350656,
        -1.81720429,  0.29156857],
       [ 2.01011032, -2.65732045, -0.03521956, ..., 10.59350656,
        -0.03542945,  0.29156857],
       ...,
       [ 0.10398314,  0.09005041, -0.33543266, ..., -0.21665953,
        -0.03542945,  0.29156857],
       [ 1.42360965,  0.09005041, -0.35877741, ..., -0.21665953,
        -0.03542945,  0.29156857],
       [-1.21564337,  0.09005041,  0.11095988, ..., -0.21665953,
        -1.65522476,  0.29156857]])

## Splitting data into Training and Test

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

# Using KERAS

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

In [ ]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

In [ ]:
# One-hot encode y_train
y_train_encoded = to_categorical(y_train)

# One-hot encode y_test (for validation)
y_test_encoded = to_categorical(y_test)

y_train_encoded

array([[0., 1.],
       [1., 0.],
       [1., 0.],
       ...,
       [0., 1.],
       [0., 1.],
       [1., 0.]])

# 1º Try

In [1]:
model = Sequential([
    Flatten(input_shape=(14,)), # input layer
    Dense(128, activation='relu'), # hidden layer 1
    Dense(128, activation='relu'), # hidden layer 2
    Dense(2, activation='softmax') # output layer
])

NameError: name 'Sequential' is not defined

In [ ]:
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

In [ ]:
type(X_train)

numpy.ndarray

In [ ]:
history = model.fit(
    X_train,
    y_train_encoded ,
    epochs=15,
    batch_size=128,
    validation_data=(X_test, y_test_encoded),
    verbose=1
)

Epoch 1/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8742 - loss: 0.2714 - val_accuracy: 0.8452 - val_loss: 0.3356
Epoch 2/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8762 - loss: 0.2659 - val_accuracy: 0.8472 - val_loss: 0.3307
Epoch 3/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8787 - loss: 0.2604 - val_accuracy: 0.8417 - val_loss: 0.3411
Epoch 4/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8789 - loss: 0.2660 - val_accuracy: 0.8449 - val_loss: 0.3357
Epoch 5/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8771 - loss: 0.2675 - val_accuracy: 0.8441 - val_loss: 0.3349
Epoch 6/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8765 - loss: 0.2664 - val_accuracy: 0.8431 - val_loss: 0.3373
Epoch 7/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8763 - loss: 0.2635 - val_accuracy: 0.8395 - val_loss: 0.3413
Epoch 8/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.8738 - loss: 0.2650 - val_accuracy: 0.

# 2º Try

In [3]:
model = Sequential([
    Flatten(input_shape=(14,)), # input layer
    Dense(256, activation='relu'), # hidden layer 1
    Dense(256, activation='relu'), # hidden layer 2
    Dense(2, activation='sigmoid') # output layer
])

NameError: name 'Sequential' is not defined

In [ ]:
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

In [ ]:
history = model.fit(
    X_train,
    y_train_encoded ,
    epochs=15,
    batch_size=128,
    validation_data=(X_test, y_test_encoded),
    verbose=1
)

Epoch 1/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - accuracy: 0.8018 - loss: 0.4082 - val_accuracy: 0.8460 - val_loss: 0.3326
Epoch 2/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8431 - loss: 0.3293 - val_accuracy: 0.8492 - val_loss: 0.3317
Epoch 3/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8521 - loss: 0.3161 - val_accuracy: 0.8482 - val_loss: 0.3251
Epoch 4/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8530 - loss: 0.3142 - val_accuracy: 0.8462 - val_loss: 0.3243
Epoch 5/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8519 - loss: 0.3165 - val_accuracy: 0.8471 - val_loss: 0.3280
Epoch 6/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8508 - loss: 0.3143 - val_accuracy: 0.8516 - val_loss: 0.3192
Epoch 7/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8558 - loss: 0.3064 - val_accuracy: 0.8452 - val_loss: 0.3235
Epoch 8/15
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8535 - loss: 0.3098 - val_accuracy: 0.

# 3º Try

In [4]:
model = Sequential([
    Flatten(input_shape=(14,)), # input layer
    Dense(256, activation='relu'), # hidden layer 1
    Dense(256, activation='relu'), # hidden layer 2
    Dense(2, activation='sigmoid') # output layer
])

NameError: name 'Sequential' is not defined

In [2]:
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

NameError: name 'model' is not defined

In [ ]:
history = model.fit(
    X_train,
    y_train_encoded ,
    epochs=30,
    batch_size=128,
    validation_data=(X_test, y_test_encoded),
    verbose=1
)

Epoch 1/30
191/191 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.8000 - loss: 0.4101 - val_accuracy: 0.8460 - val_loss: 0.3324
Epoch 2/30
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8477 - loss: 0.3272 - val_accuracy: 0.8462 - val_loss: 0.3278
Epoch 3/30
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8499 - loss: 0.3178 - val_accuracy: 0.8468 - val_loss: 0.3249
Epoch 4/30
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8515 - loss: 0.3188 - val_accuracy: 0.8512 - val_loss: 0.3258
Epoch 5/30
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8521 - loss: 0.3150 - val_accuracy: 0.8482 - val_loss: 0.3238
Epoch 6/30
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - accuracy: 0.8547 - loss: 0.3085 - val_accuracy: 0.8468 - val_loss: 0.3287
Epoch 7/30
191/191 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.8531 - loss: 0.3108 - val_accuracy: 0.8469 - val_loss: 0.3244
Epoch 8/30
191/191 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8555 - loss: 0.3110 - val_accuracy: 0.

# Conclusion

O dataset escolhido possui 13 features e uma classe (<50K ou >=50K). O objetivo foi avaliar se dado input, contendo informações sobre trabalho, estado civil, educação, sexo etc. ganha mais ou menos de 50K por mês.

Utilizando o Keras em conjunto com o Tensorflow e diferentes parametros para o modelo, foi constatado que a acuracia da rede foi de valores em torno de 86 a 88% de acerto. Foram realizados diferentes testes com três variações possíveis dos parametros de simulação.

No primeiro teste (1ª tentativa) foi utilizado os mesmoos parametros do exemplo do professor, mas modificado para o dataset utilizado

```
model = Sequential([
    Flatten(input_shape=(14,)), # input layer
    Dense(128, activation='relu'), # hidden layer 1
    Dense(128, activation='relu'), # hidden layer 2
    Dense(2, activation='softmax') # output layer
])

model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

X_train,
    y_train_encoded ,
    epochs=15,
    batch_size=128,
    validation_data=(X_test, y_test_encoded),
    verbose=1
```

No segundo teste, foi utilizada uma quantidade maior de neuronios na camada oculta (de 128 p/ 256) e a função de ativação na saida foi alterada para "sigmoid".

```
model = Sequential([
    Flatten(input_shape=(14,)), # input layer
    Dense(256, activation='relu'), # hidden layer 1
    Dense(256, activation='relu'), # hidden layer 2
    Dense(2, activation='sigmoid') # output layer
])

model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

history = model.fit(
    X_train,
    y_train_encoded ,
    epochs=15,
    batch_size=128,
    validation_data=(X_test, y_test_encoded),
    verbose=1
)
```

E, por fim, no terceiro teste, a única alteração foi o número de épocas em relação ao teste 2, de 15 p/ 30.


```
model = Sequential([
    Flatten(input_shape=(14,)), # input layer
    Dense(256, activation='relu'), # hidden layer 1
    Dense(256, activation='relu'), # hidden layer 2
    Dense(2, activation='sigmoid') # output layer
])
```
A conclusão que é possível tirar desse exercício é que a quantidade de neurõnios nas hidden layers e um número de épocas baixo no primeiro teste teve um desempenho melhor que no segundo, em que o número de neuronios foi acrescido mas a acurácia diminuiu. Já no terceiro teste é possível ver uma acurácia maior que no primeiro, mas a diferença foi mínima em relação ao poder computacional utilizado, já que o número de épocas foi dobrado, além do número de neuronios. Portanto, o primeiro teste é o mais eficiente.
